google/auth/credentials.py

```diff
@@ -0,0 +1,209 @@
+    construction. Some classes will provide mechanisms to copy the credentials
+    with modifications such as :meth:`ScopedCredentials.with_scopes`.
+    """
+    def __init__(self):
+        self.token = None
+        """str: The bearer token that can be used in HTTP headers to make
+        authenticated requests."""
+        self.expiry = None
+        """Optional[datetime]: When the token expires and is no longer valid.
+        If this is None, the token is assumed to never expire."""
+
+    @property
+    def valid(self):
+        """Checks the validity of the credentials.
+
+        This is True if the credentials have a :attr:`token` and the token
+        is not :attr:`expired`.
+        """
+        return self.token is not None and not self.expired
+
+    @property
+    def expired(self):
+        """Checks if the credentials are expired.
+
+        Note that credentials can be invalid but not expired becaue Credentials
+        with :attr:`expiry` set to None is considered to never expire.
+        """
+        now = _helpers.utcnow()
+        if self.expiry is None or self.expiry > now:
+            return False
+        else:
+            return True
+
+    @abc.abstractmethod
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # pylint: disable=missing-raises-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Refresh must be implemented')
+
+    def apply(self, headers, token=None):
+        """Apply the token to the authentication header.
+
+        Args:
+            headers (Mapping): The HTTP request headers.
+            token (Optional[str]): If specified, overrides the current access
+                token.
+        """
+        headers[b'authorization'] = 'Bearer {}'.format(
+            _helpers.from_bytes(token or self.token))
+
+    def before_request(self, request, method, url, headers):
+        """Performs credential-specific before request logic.
+
+        Refreshes the credentials if necessary, then calls :meth:`apply` to
+        apply the token to the authentication header.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
+        # (Subclasses may use these arguments to ascertain information about
+        # the http request.)
+        if not self.valid:
+            self.refresh(request)
+        self.apply(headers)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ScopedCredentials(object):
```

Comment:
Regarding the MRO: that hadn't occurred to me; it's been so long since I wrote anything that has one concrete method implementation override another concrete method implementation. I suppose I forgot that people did that. It makes code shorter, but doesn't it make it harder to reason about?

Regarding separation without much purpose: yes, it's separation without much purpose _now_. The purpose is to remain able to respond to the needs of the future by having made minimal promises in the past. That's why I'm advocating for all this extra consideration only on the API surface of the library; it is more code and more abstraction and there would be no point internally where everything remains free for all time to be changed in every way.

Regarding how classes are presented: yes, it is an abstraction violation to have a public class inherit from a private class. Why have concrete classes in the API at all? If you already have an interface that describes the full behavior that the object passed from the library to the application will support, why not just have a function that returns an instance of that interface? Something like
```py
class Credentials:
  # some public abstract method specifications

class ScopedCredentials(Credentials):
  # some more public abstract method specifications

def fixed_scoped_credentials(token, scopes):
  """<summary string>

  Args:
    <parameter descriptions>

  Returns:
    A ScopedCredentials that <more information>.
  """
  import _private_implementation_module
  return _private_implementation_module.FixedScopedCredentials(token, scopes)
```
? Including a public class in an API is a big commitment - much bigger than including an interface describing what the API user may do with a type and one or more functions available for constructing objects that implement the type. Does that big commitment need to be made? Or is it just so tantalizingly straightforward and direct?

You may be worried about the situation of having presented several interfaces and needing to pass out of the library objects that implement several of the interfaces - it would be truthful and complete to say "just describe the value as an instance of all the interfaces it implements" but in practice I've met friction doing this. An alternative is to define extra interfaces that define no methods of their own but by inheriting from two or more others serve to specify a union of them - in theory this leads to exponential empty interface writing but in practice it leads to one or two. Will `google-auth-library-python` emit any object that is a `ScopedCredentials` but not also a `Credentials`? If not, should `ScopedCredentials` inherit from `Credentials`? Will the library emit any object that is a `SigningCredentials` but not also a `Credentials`? If not, should `SigningCredentials` be a subinterface of `Credentials`? Will the library emit any value that is both a `SigningCredentials` and a `ScopedCredentials`? If so, should a `ComboCredentials(ScopedCredentials, SigningCredentials)` be defined?

Sorry that I'm dragging this out and probably not advocating well enough. Years of experience has led me to believe that these questions of interface design are crucially important, and I'm always amazed by how few other people's experiences lead them to the same perspective. Maybe I've made more mistakes along the way than I've realized...